### ISSUE TITLE
Extraneous -Wshadow diagnostic for local variable redefinition

### ISSUE THREAD

**Issue by llvmbot** [llvm/llvm-project#13776]
|  |  |
| --- | --- |
| Bugzilla Link | [13404](https://llvm.org/bz13404) |
| Version | trunk |
| OS | All |
| Reporter | LLVM Bugzilla Contributor |
| CC | @DougGregor |

## Extended Description 
The code:
```c++
struct S {};

int main() {
    S S;
    struct S S;
}
```

results in two diagnostics:
```
ConsoleApplication1.cpp:5:14: warning: declaration shadows a local variable [-Wshadow]
    struct S S;
             ^
ConsoleApplication1.cpp:4:7: note: previous declaration is here
    S S;
      ^
ConsoleApplication1.cpp:5:14: error: redefinition of 'S'
    struct S S;
             ^
ConsoleApplication1.cpp:4:7: note: previous definition is here
    S S;
      ^
```
The -Wshadow diagnostic should not trigger when the declaration is in fact an illegal redefinition.

**Comment by philnik777** [comment]
Fixed in Clang 4.0: https://godbolt.org/z/e76Kqn134
